src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs

```diff
@@ -164,93 +167,112 @@ private bool CheckGlobalConditions(BuildUpToDateCheckLogger logger, State state)
             return true;
         }
 
-        /// <summary>
-        /// Returns one of:
-        /// <list type="bullet">
-        ///     <item><c>(time, path)</c> describing the earliest output when all were found</item>
-        ///     <item><c>(null, path)</c> where <c>path</c> is the first output that could not be found</item>
-        ///     <item><c>(null, null)</c> when there were no outputs</item>
-        /// </list>
-        /// </summary>
-        private static (DateTime? time, string? path) GetEarliestOutput(IEnumerable<string> outputs, in TimestampCache timestampCache)
+        private bool CheckInputsAndOutputs(BuildUpToDateCheckLogger logger, in TimestampCache timestampCache, State state)
         {
-            DateTime? earliest = DateTime.MaxValue;
-            string? earliestPath = null;
-            bool hasOutput = false;
+            // UpToDateCheckInput/Output/Built items have optional 'Set' metadata that determine whether they
+            // are treated separately or not. If omitted, such inputs/outputs are included in the default set,
+            // which also includes other items such as project files, compilation items, analyzer references, etc.
 
-            foreach (string output in outputs)
+            // First, validate the relationship between inputs and outputs within the default set.
+            if (!CheckInputsAndOutputs(CollectDefaultInputs(), CollectDefaultOutputs(), timestampCache, DefaultSetName))
```

Comment:
Why do the default set separately verses just having a special case in the log messages, which is already happening in some log messages anyway? At a quick glance it looks like you could just not remove `DefaultSetName` from `SetNames` and everything would work as expected, without needing to duplicate the CollectXX() methods?

Reply:
The default sets contain a bunch more items that are subject to comparisons, but not defined as `BuildUpToDateCheck*` items. Have a look at the implementation of `CollectDefaultInputs`, for example.

This went through several iterations to try and remove redundancy. There may be more opportunity here, but I've stared at this for a while and can't see it.